Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,816 @@
+  }
+
+  public static func |= (lhs: inout BigInt, rhs: BigInt) {
+    var rhsWords = rhs.words
+    BigInt._signExtend(lhsWords: &lhs.words, rhsWords: &rhsWords)
+
+    for i in 0 ..< rhsWords.count {
+      lhs.words[i] |= rhsWords[i]
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  public static func ^= (lhs: inout BigInt, rhs: BigInt) {
+    var rhsWords = rhs.words
+    BigInt._signExtend(lhsWords: &lhs.words, rhsWords: &rhsWords)
+
+    for i in 0 ..< rhsWords.count {
+      lhs.words[i] &= rhsWords[i]
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  public static func <<= <RHS>(lhs: inout BigInt, rhs: RHS) where RHS: BinaryInteger {
+    if rhs.signum() < 0 {
+      lhs >>= rhs.magnitude
+      return
+    }
+
+    let wordLength = UInt.bitWidth
+    let isNegative = lhs._isNegative
+
+    let (fullWords, remainder) = rhs.quotientAndRemainder(dividingBy: RHS(wordLength))
+    lhs.words = Words(repeating: 0, count: Int(fullWords)) + lhs.words + [isNegative ? UInt.max : 0]
+
+    if remainder > 0 {
+      var value: UInt = 0
+      for i in 0 ..< lhs.words.count {
+        let temp = lhs.words[i] >> (wordLength - Int(remainder))
+        lhs.words[i] <<= Int(remainder)
+        if i > 0 {
+          lhs.words[i] &= UInt.max << Int(remainder)
+          lhs.words[i] |= value
+        }
+
+        value = temp
+      }
+
+      if isNegative {
+        lhs.words[lhs.words.count - 1] |= (UInt.max << Int(remainder))
+      }
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  public static func >>= <RHS>(lhs: inout BigInt, rhs: RHS) where RHS: BinaryInteger {
+    if rhs.signum() < 0 {
+      lhs <<= rhs.magnitude
+      return
+    }
+
+    let wordLength = UInt.bitWidth
+    let isNegative = lhs._isNegative
+
+    let (fullWords, remainder) = rhs.quotientAndRemainder(dividingBy: RHS(wordLength))
+    if fullWords < lhs.words.count {
+      lhs.words.removeFirst(Int(fullWords))
+    } else {
+      lhs = isNegative ? -1 : 0
+      return
+    }
+
+    if remainder > 0 {
+      let mask = ~(UInt.max << remainder)
+      var value: UInt = 0
+      for i in stride(from: lhs.words.count - 1, through: 0, by: -1) {
+        let temp = lhs.words[i] & mask
+        lhs.words[i] >>= remainder
+        lhs.words[i] |= (value << (UInt.bitWidth - Int(remainder)))
+        value = temp
+      }
+    }
+
+    if isNegative {
+      lhs.words[lhs.words.count - 1] |= (UInt.max << (UInt.bitWidth - Int(remainder)))
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  @inlinable
+  public func quotientAndRemainder(dividingBy rhs: BigInt) -> (quotient: BigInt, remainder: BigInt) {
+    return BigInt._div(lhs: self, rhs: rhs)
+  }
+
+  @inlinable
+  public func signum() -> BigInt {
+    if _isNegative {
+      return -1
+    } else if self == 0 {
+      return 0
+    }
+
+    return 1
+  }
+}
+
+// MARK: -
+
+extension BigInt {
+
+  private static func _findQhat(
```

Comment:
It would be a good idea to add an explicit reference to TAOCP so that people can figure out how this works.

Reply:
That is a good idea, yes.